components/post_view/reaction/reaction.jsx

```diff
@@ -10,88 +10,152 @@ import OverlayTrigger from 'components/overlay_trigger';
+                displayNumber: reactionCount,
+                reactedNumber: reactionCount + 1,
+                unreactedNumber: reactionCount,
+            };
+        }
+
+        if (currentUserReacted && !canRemoveReaction) {
+            this.state = {
+                ...this.state,
+                reactedClass: 'Reaction read-only',
+            };
+        } else if (!currentUserReacted && !canAddReaction) {
+            this.state = {
+                ...this.state,
+                reactedClass: 'Reaction read-only',
+            };
+        }
+
+        this.reactionButtonRef = React.createRef();
+        this.reactionCountRef = React.createRef();
+        this.reactedNumeralRef = React.createRef();
     }
 
-    handleRemoveReaction = (e) => {
-        e.preventDefault();
-        this.props.actions.removeReaction(this.props.post.id, this.props.emojiName);
+    shouldComponentUpdate(nextProps) {
+        // reaction count is changing
+        if (nextProps.reactionCount !== this.props.reactionCount) {
+            // if current user is reacting
+            if (!this.props.sortedUsers.currentUserReacted && nextProps.sortedUsers.currentUserReacted) {
+                this.setState({
+                    displayNumber: nextProps.reactionCount,
+                    reactedNumber: nextProps.reactionCount,
+                    unreactedNumber: nextProps.reactionCount - 1,
+                });
+                return true;
+            }
+
+            // if current user is unreacting
+            if (this.props.sortedUsers.currentUserReacted && !nextProps.sortedUsers.currentUserReacted) {
+                this.setState({
+                    displayNumber: nextProps.reactionCount,
+                    reactedNumber: nextProps.reactionCount + 1,
+                    unreactedNumber: nextProps.reactionCount,
+                });
+                return true;
+            }
+
+            // if another user is reacting/unreacting
+            if (this.props.sortedUsers.currentUserReacted === nextProps.sortedUsers.currentUserReacted) {
+                if (this.props.sortedUsers.currentUserReacted) {
+                    this.setState({
+                        displayNumber: nextProps.reactionCount,
+                        reactedNumber: nextProps.reactionCount,
+                        unreactedNumber: nextProps.reactionCount - 1,
+                    });
+                    return true;
+                }
+                this.setState({
+                    displayNumber: nextProps.reactionCount,
+                    reactedNumber: nextProps.reactionCount + 1,
+                    unreactedNumber: nextProps.reactionCount,
+                });
+                return true;
+            }
+        }
+        return true;
+    }
+
+    handleClick = () => {
+        // only proceed if user has permission to react
+        if (!this.props.canAddReaction || !this.props.canRemoveReaction) {
```

Comment:
this can be simplified `!(canAddReaction && canRemoveReaction)`

Reply:
Yep, thanks.